File: ta4j-core/src/main/java/org/ta4j/core/indicators/pivotpoints/StandardReversalIndicator.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.indicators.pivotpoints;

import static org.ta4j.core.num.NaN.NaN;

import java.util.List;

import org.ta4j.core.Bar;
import org.ta4j.core.indicators.RecursiveCachedIndicator;
import org.ta4j.core.num.Num;

/**
 * Pivot Reversal Indicator.
 *
 * @see <a href=
 *      "https://chartschool.stockcharts.com/table-of-contents/technical-indicators-and-overlays/technical-overlays/pivot-points">
 *      https://chartschool.stockcharts.com/table-of-contents/technical-indicators-and-overlays/technical-overlays/pivot-points</a>
 */
public class StandardReversalIndicator extends RecursiveCachedIndicator<Num> {

    private final PivotPointIndicator pivotPointIndicator;
    private final PivotLevel level;
    private final Num two;

    /**
     * Constructor.
     *
     * Calculates the (standard) reversal for the corresponding pivot level
     *
     * @param pivotPointIndicator the {@link PivotPointIndicator} for this reversal
     * @param level               the {@link PivotLevel} for this reversal
     */
    public StandardReversalIndicator(PivotPointIndicator pivotPointIndicator, PivotLevel level) {
        super(pivotPointIndicator);
        this.pivotPointIndicator = pivotPointIndicator;
        this.level = level;
        this.two = pivotPointIndicator.getBarSeries().numFactory().numOf(2);
    }

    @Override
    protected Num calculate(int index) {
        List<Integer> barsOfPreviousPeriod = pivotPointIndicator.getBarsOfPreviousPeriod(index);
        if (barsOfPreviousPeriod.isEmpty()) {
            return NaN;
        }
        switch (level) {
        case RESISTANCE_3:
            return calculateR3(barsOfPreviousPeriod, index);
        case RESISTANCE_2:
            return calculateR2(barsOfPreviousPeriod, index);
        case RESISTANCE_1:
            return calculateR1(barsOfPreviousPeriod, index);
        case SUPPORT_1:
            return calculateS1(barsOfPreviousPeriod, index);
        case SUPPORT_2:
            return calculateS2(barsOfPreviousPeriod, index);
        case SUPPORT_3:
            return calculateS3(barsOfPreviousPeriod, index);
        default:
            return NaN;
        }
    }

    @Override
    public int getCountOfUnstableBars() {
        return 0;
    }

    private Num calculateR3(List<Integer> barsOfPreviousPeriod, int index) {
        Bar bar = getBarSeries().getBar(barsOfPreviousPeriod.get(0));
        Num low = bar.getLowPrice();
        Num high = bar.getHighPrice();
        for (int i : barsOfPreviousPeriod) {
            Bar iBar = getBarSeries().getBar(i);
            low = iBar.getLowPrice().min(low);
            high = iBar.getHighPrice().max(high);
        }
        return high.plus(two.multipliedBy((pivotPointIndicator.getValue(index).minus(low))));
    }

    private Num calculateR2(List<Integer> barsOfPreviousPeriod, int index) {
        Bar bar = getBarSeries().getBar(barsOfPreviousPeriod.get(0));
        Num low = bar.getLowPrice();
        Num high = bar.getHighPrice();
        for (int i : barsOfPreviousPeriod) {
            Bar iBar = getBarSeries().getBar(i);
            low = iBar.getLowPrice().min(low);
            high = iBar.getHighPrice().max(high);
        }
        return pivotPointIndicator.getValue(index).plus((high.minus(low)));
    }

    private Num calculateR1(List<Integer> barsOfPreviousPeriod, int index) {
        Num low = getBarSeries().getBar(barsOfPreviousPeriod.get(0)).getLowPrice();
        for (int i : barsOfPreviousPeriod) {
            low = (getBarSeries().getBar(i).getLowPrice()).min(low);
        }
        return two.multipliedBy(pivotPointIndicator.getValue(index)).minus(low);
    }

    private Num calculateS1(List<Integer> barsOfPreviousPeriod, int index) {
        Num high = getBarSeries().getBar(barsOfPreviousPeriod.get(0)).getHighPrice();
        for (int i : barsOfPreviousPeriod) {
            high = (getBarSeries().getBar(i).getHighPrice()).max(high);
        }
        return two.multipliedBy(pivotPointIndicator.getValue(index)).minus(high);
    }

    private Num calculateS2(List<Integer> barsOfPreviousPeriod, int index) {
        Bar bar = getBarSeries().getBar(barsOfPreviousPeriod.get(0));
        Num high = bar.getHighPrice();
        Num low = bar.getLowPrice();
        for (int i : barsOfPreviousPeriod) {
            Bar iBar = getBarSeries().getBar(i);
            high = iBar.getHighPrice().max(high);
            low = iBar.getLowPrice().min(low);
        }
        return pivotPointIndicator.getValue(index).minus((high.minus(low)));
    }

    private Num calculateS3(List<Integer> barsOfPreviousPeriod, int index) {
        Bar bar = getBarSeries().getBar(barsOfPreviousPeriod.get(0));
        Num high = bar.getHighPrice();
        Num low = bar.getLowPrice();
        for (int i : barsOfPreviousPeriod) {
            Bar iBar = getBarSeries().getBar(i);
            high = iBar.getHighPrice().max(high);
            low = iBar.getLowPrice().min(low);
        }
        return low.minus(two.multipliedBy((high.minus(pivotPointIndicator.getValue(index)))));
    }
}

File: ta4j-core/src/main/kotlin/org/ta4j/core/indicators/pivotpoints/StandardReversalIndicator.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.indicators.pivotpoints

import org.ta4j.core.indicators.RecursiveCachedIndicator
import org.ta4j.core.num.NaN
import org.ta4j.core.num.*

/**
 * Pivot Reversal Indicator.
 *
 * @see [
 * http://stockcharts.com/school/doku.php?id=chart_school:technical_indicators:pivot_points](http://stockcharts.com/school/doku.php?id=chart_school:technical_indicators:pivot_points)
 */
/**
 * Constructor.
 *
 * Calculates the (standard) reversal for the corresponding pivot level
 *
 * @param pivotPointIndicator the [PivotPointIndicator] for this reversal
 * @param level               the [PivotLevel] for this reversal
 */
class StandardReversalIndicator(private val pivotPointIndicator: PivotPointIndicator, private val level: PivotLevel) :
    RecursiveCachedIndicator<Num>(
        pivotPointIndicator
    ) {
    override fun calculate(index: Int): Num {
        val barsOfPreviousPeriod = pivotPointIndicator.getBarsOfPreviousPeriod(index)
        return if (barsOfPreviousPeriod.isEmpty()) {
            NaN.NaN
        } else when (level) {
            PivotLevel.RESISTANCE_3 -> calculateR3(barsOfPreviousPeriod, index)
            PivotLevel.RESISTANCE_2 -> calculateR2(barsOfPreviousPeriod, index)
            PivotLevel.RESISTANCE_1 -> calculateR1(barsOfPreviousPeriod, index)
            PivotLevel.SUPPORT_1 -> calculateS1(barsOfPreviousPeriod, index)
            PivotLevel.SUPPORT_2 -> calculateS2(barsOfPreviousPeriod, index)
            PivotLevel.SUPPORT_3 -> calculateS3(barsOfPreviousPeriod, index)
        }
    }

    private fun calculateR3(barsOfPreviousPeriod: List<Int>, index: Int): Num {
        val bar = barSeries!!.getBar(barsOfPreviousPeriod[0])
        var low = bar.lowPrice!!
        var high = bar.highPrice!!
        for (i in barsOfPreviousPeriod) {
            low = barSeries.getBar(i).lowPrice!!.min(low)
            high = barSeries.getBar(i).highPrice!!.max(high)
        }
        return high + numOf(2) * (pivotPointIndicator[index] - low)
    }

    private fun calculateR2(barsOfPreviousPeriod: List<Int>, index: Int): Num {
        val bar = barSeries!!.getBar(barsOfPreviousPeriod[0])
        var low = bar.lowPrice!!
        var high = bar.highPrice!!
        for (i in barsOfPreviousPeriod) {
            low = barSeries.getBar(i).lowPrice!!.min(low)
            high = barSeries.getBar(i).highPrice!!.max(high)
        }
        return pivotPointIndicator[index] + (high - low)
    }

    private fun calculateR1(barsOfPreviousPeriod: List<Int>, index: Int): Num {
        var low = barSeries!!.getBar(barsOfPreviousPeriod[0]).lowPrice!!
        for (i in barsOfPreviousPeriod) {
            low = barSeries.getBar(i).lowPrice!!.min(low)
        }
        return (numOf(2) * pivotPointIndicator[index]) - low
    }

    private fun calculateS1(barsOfPreviousPeriod: List<Int>, index: Int): Num {
        var high = barSeries!!.getBar(barsOfPreviousPeriod[0]).highPrice!!
        for (i in barsOfPreviousPeriod) {
            high = barSeries.getBar(i).highPrice!!.max(high)
        }
        return numOf(2) * pivotPointIndicator[index] - high
    }

    private fun calculateS2(barsOfPreviousPeriod: List<Int>, index: Int): Num {
        val bar = barSeries!!.getBar(barsOfPreviousPeriod[0])
        var high = bar.highPrice!!
        var low = bar.lowPrice!!
        for (i in barsOfPreviousPeriod) {
            high = barSeries.getBar(i).highPrice!!.max(high)
            low = barSeries.getBar(i).lowPrice!!.min(low)
        }
        return pivotPointIndicator[index] - (high - low)
    }

    private fun calculateS3(barsOfPreviousPeriod: List<Int>, index: Int): Num {
        val bar = barSeries!!.getBar(barsOfPreviousPeriod[0])
        var high = bar.highPrice!!
        var low = bar.lowPrice!!
        for (i in barsOfPreviousPeriod) {
            high = barSeries.getBar(i).highPrice!!.max(high)
            low = barSeries.getBar(i).lowPrice!!.min(low)
        }
        return low - numOf(2) * (high - pivotPointIndicator[index])
    }
}